internal/lsp/connection.go

```diff
@@ -0,0 +1,224 @@
+	ExcludeMethods []string
+}
+
+func (cfg *ConnectionLoggingConfig) ShouldLog(method string) bool {
+	if len(cfg.IncludeMethods) > 0 {
+		for _, m := range cfg.IncludeMethods {
+			if m == method {
+				return true
+			}
+		}
+
+		return false
+	}
+
+	for _, m := range cfg.ExcludeMethods {
+		if m == method {
+			return false
+		}
+	}
+
+	return true
+}
+
+type ConnectionHandlerFunc func(context.Context, *jsonrpc2.Conn, *jsonrpc2.Request) (result interface{}, err error)
+
+type connectionLogger struct {
+	writer io.Writer
+}
+
+func (c *connectionLogger) Printf(format string, v ...interface{}) {
+	fmt.Fprintf(c.writer, format, v...)
+}
+
+func NewConnectionFromLanguageServer(
+	ctx context.Context,
+	handler ConnectionHandlerFunc,
+	opts *ConnectionOptions,
+) *jsonrpc2.Conn {
+	return jsonrpc2.NewConn(
+		ctx,
+		jsonrpc2.NewBufferedStream(StdOutReadWriteCloser{}, jsonrpc2.VSCodeObjectCodec{}),
+		jsonrpc2.HandlerWithError(handler),
+		logMessages(opts.LoggingConfig),
+	)
+}
+
+func logMessages(cfg ConnectionLoggingConfig) jsonrpc2.ConnOpt {
+	logger := &connectionLogger{writer: cfg.Writer}
+
+	return func(c *jsonrpc2.Conn) {
+		// Remember reqs we have received so that we can helpfully show the
+		// request method in OnSend for responses.
+		var (
+			mu         sync.Mutex
+			reqMethods = map[jsonrpc2.ID]string{}
+		)
+
+		setMethod := func(id jsonrpc2.ID, method string) {
+			mu.Lock()
+			defer mu.Unlock()
+
+			reqMethods[id] = method
+		}
+
+		getMethod := func(id jsonrpc2.ID) string {
+			mu.Lock()
+			defer mu.Unlock()
+
+			return reqMethods[id]
+		}
+
+		deleteMethod := func(id jsonrpc2.ID) {
+			mu.Lock()
+			defer mu.Unlock()
+
+			delete(reqMethods, id)
+		}
+
+		if cfg.LogInbound {
+			jsonrpc2.OnRecv(buildRecvHandler(setMethod, logger, cfg))(c)
+		}
+
+		if cfg.LogOutbound {
+			jsonrpc2.OnSend(buildSendHandler(getMethod, deleteMethod, logger, cfg))(c)
+		}
+	}
+}
+
+func buildRecvHandler(
+	setMethod func(jsonrpc2.ID, string),
+	logger *connectionLogger,
+	cfg ConnectionLoggingConfig,
+) func(req *jsonrpc2.Request, resp *jsonrpc2.Response) {
+	return func(req *jsonrpc2.Request, resp *jsonrpc2.Response) {
+		switch {
+		case req != nil && resp == nil:
+			setMethod(req.ID, req.Method)
+
+			if !cfg.ShouldLog(req.Method) {
+				return
+			}
+
+			params, _ := json.Marshal(req.Params)
+			if req.Notif {
+				logger.Printf("--> notif: %s: %s\n", req.Method, params)
+			} else {
+				logger.Printf("--> request #%s: %s: %s\n", req.ID, req.Method, params)
+			}
+
+		case resp != nil:
+			var method string
+			if req != nil {
+				method = req.Method
+			} else {
+				method = "(no matching request)"
+			}
+
+			if !cfg.ShouldLog(method) {
+				return
+			}
+
+			switch {
+			case resp.Result != nil:
+				result, _ := json.Marshal(resp.Result)
+				logger.Printf("--> response #%s: %s: %s\n", resp.ID, method, result)
+			case resp.Error != nil:
+				err, _ := json.Marshal(resp.Error)
```

Comment:
Heh, using `err` for anything but an `Error` sure is something 😅

Reply:
Yeah... I have altered that now... 9892c609e05ea8ad441f5ff987ef8a38e645b2eb